### PR TITLE
[CNSMR-2044] Update Release Process for uniformisation of branches & tags

### PR DESCRIPTION
**Pull request by AliSoftware**
# Why

In preparation for updating our automation of the release process, especially for automating the computation of the CHANGELOG from the commit lists and the creation of the CRP ticket that will include this auto-built CHANGELOG, we need to first standardise the way we name our release branches and tags, so that the bot can then find them in a reliable way and make assumptions on their content (extract which app the tag applies to, etc)

_Note that this naming convention will also be applied to our colleagues in the Android chapter, in order to also have unified naming of release branches and tags between Android and iOS repos too._

# How

The goal here is to enforce the following conventions on branches and tags naming:

| Product to Release | Branch Name | Tag Name once published |
|---|---|---|
| Release for Babylon | `release/babylon/1.2.3` | `babylon/1.2.3` |
| Release for Telus | `release/telus/1.2.3` | `telus/1.2.3` |
| Release for Bupa | `release/bupa/1.2.3` | `bupa/1.2.3` |
| Release for the SDK | `release/sdk/0.1.2` | `sdk/0.2.1` |
| Release for a public SDK beta | `release/sdk/0.1.2-beta1` | `sdk/0.2.1-beta1` |

Note that we'll not have branches named `release/x.y.z` from now on, and instead using `release/babylon/x.y.z`, so that every white-label app follow the same convention and the main Babylon target won't be an exception.

Also note that we'll now have **separate branches tags for Babylon and Telus**. This will allow us to create **separate GitHub Releases** – with potentially different descriptions (= AppStore release notes) and each with their respective `.ipa` uploaded in artefacts – which is something we failed to do until now.

## CRP command logic for creating CRP tickets and CHANGELOG

The CRP command will deduce which apps the release is for using the release branch name (e.g. it will create a CRP ticket for Telus if `/crp ios branch:release/telus/1.2.3`)

It will also automatically find the tag corresponding to the previous release (e.g. tag `telus/1.2.2` if we're creating the CRP for `release/telus/1.2.3`), hence the importance of respecting that new naming convention and having a tag for each separate product.

### CODE REVIEW

**Comment by AliSoftware** [comment]
Note: this PR is pending validation of the naming convention – and the CRP/CHANGELOG auto-build [algorithm described in this Google Doc]() – by both iOS and Android teams.

Since we finally managed to find a solution that avoids most of the risks and tricky cases that could happen (ensuring the CHANGELOG for Babylon is not considering the tag for Telus when searching for the previous release) while avoiding compliexifying the process for devs (avoiding to have to have separate release branches for Babylon and Telus), I'd say it's now in a shape likely to be approved, so the updates to Release Process in this PR are likely to be what will be approved, but until we have the official OK from all parties we should wait to merge this.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (dmcrodrigues) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: You mean _and push it to_?

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (dmcrodrigues) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: ❤️

**Review by dmcrodrigues — APPROVED** [review]

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: Note: There are plans to enforce that convention via Danger – so that any PR which contains changes to any fiiles in the `SDK/` folder will be required to have the `#SDK` tag in its title to ensure the squashed commit will be listed in the SDK CRP ticket.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: @dmcrodrigues We should actually suggest to do those two actions (TestFlight build and Hockey Build) for both apps (Babylon AND Telus), right?

Comment (dmcrodrigues) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: I think @RuiAAPeres and @chitrakotwani are the best persons to answer this because I’m not sure if there’s any coordination/approval with Telus for the releases.

Comment (chitrakotwani) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: @AliSoftware TL; DR - No

Telus branch is cut from Babylon release branch(`release/x.y.z`) only after it is signed off for app-store. This is so that Telus is forked from a fully tested and stable Babylon branch. 

Then,  testing team only tests tickets related to Telus on Telus release branch before signing off Telus build.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]:
```suggestion
  * For releases specific to apps other than main Babylon app (Bupa, NHS111, Telus ...), name it `release/{appname}/{version}` (e.g. `release/bupa/3.2.0`)
```

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: I think this document can benefit from a section specific to Telus. This knowledge is buried in emails at the moment. Let me do it after your PR is merged.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]:
```suggestion
  * This creates a TestFlight build (try to make one as early as possible so that you can catch issues like missing/expired certificates or profiles and any other production build errors early).
```

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: I can see Telus being mentioned alongside Babylon everywhere, how does it work for other apps like NHS111?  It seems Telus should be mentioned alongside those apps. 

Since release engineer typically does not handle Telus releases, it could potentially cause confusion. 

Also, Telus does not release every time main Babylon app is released

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: I think this as well we can specify in a section dedicated to Telus

**Review by chitrakotwani — CHANGES_REQUESTED** [review]

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: Wait that's not the info I had, and changes the algorithm and the convention quite a bit then 😅

We should then not use a common `release/x.y.z` at all anymore then (and the CRP should not consider creating two CRP tickets from that single branch name) and would instead only use `release/{appname}/{version}` for all cases.

I'll loop back that info in the Google doc since reviewers didn't seem to be aware and that changes the algorithm of we always end up having two branches with different names for Babylon and Telus in all cases, right?

Comment (chitrakotwani) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: Which Google doc are you referring to? 

Telus and Babylon always have different branch names - that's correct

Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: Sorry, I'm referring to the Google Doc I mentioned in https://github.com/Babylonpartners/ios-playbook/pull/158#issuecomment-501385077 which I'm working on (and aims to define the the algorithm for the CRP automation while considering all the complexities of our mono-repo-but-multiple-products config)

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: My original (erroneous) understanding was that Telus and Babylon were released in tandem, and I thus assumed that it was done during the same cycle and by the same release engineer, doing both at the same time, and from the same branch; hence the mention of a common `release/x.y.z` branch for both, and the joint mention. While for Bupa/NHS111/... I already noted it was totally out of the release cycle and separate cases.

After gathering some more info (especially from you 🙇) it appears my assumption was wrong and that Telus is in fact to be considered a separate release process by a separate release engineer, so no different from Bupa / NHS111 / ... (even if we likely release Telus more often than Bupa/NHS111).

I'll amend the document with this new information to propose instead that:

 - Since we don't need a special case with a common branch from which multiple apps (Babylon+Telus) would be published, but instead have separate release branches even for Telus (whose release branch is cut off of the Babylon's release branch once QA'd), I'll remove the case of the `release/x.y.z` branch name
 - **This mean we'll not use the `release/x.y.z` name anymore, and instead always use `release/{app}/{version}`** naming – even for Babylon and Telus

So a typical release cycle which happen to plan to both release Babylon and Telus for a version 1.2.3 would mean creating a `release/babylon/1.2.3` release branch first, wait for QA to sign it off, then create the `release/telus/1.2.3` branch from the `release/babylon/1.2.3` branch and make QA sign off the Telus-specific additions.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AnnKatF) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: Shouldn't we mention in which channel should this command be triggered?

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AnnKatF) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: Can we mention who is responsible for creating the release JIRA board? (in the past was the release engineer responsibility)

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AnnKatF) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: Shouldn't we have more clarity on what is involved in "compliance info" submission?

**Review by AnnKatF — COMMENTED** [review]
This is a very easy to follow document for any new release engineer.
Good work 👍

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: According to [the more general doc in Confluence](https://babylonpartners.atlassian.net/wiki/spaces/V4/pages/767623303/iOS+Android+Release+process#iOS&AndroidReleaseprocess-JiraReleaseBoard) it's Lilla's responsibility – that's also why it's not listed here as part of the Release Engineer's tasks I guess

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: That's a good question (you'll probably need to ask Andreea or Lilla for more info)

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: So, you chatted with Lilla and in fact the confluence do is not up-to-date, but it's still not the Release Engineer's role to update it.

> [Lilla] The release board does exist, it’s just the filter that needs to be changed. Which is done by @JasonDobo

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: ℹ️ There are future plans to enforce that convention via Danger – so that any PR which contains changes to any files in the SDK/ folder will be required to have the `#SDK` tag in its title to ensure the squashed commit will be listed in the SDK CRP ticket. I'll do a PR to add that Danger rule soon.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: ℹ️ This currently needs [a fix in our Steve bot so that it takes the new branch naming convention into account](https://github.com/Babylonpartners/ios-build-distribution/blob/master/Sources/AppCore/Models/Constants.swift#L47).

I'll do a PR in `ios-build-distribute` for this soon.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: ℹ️ Yet to be implemented in a future PR in Stevenson (the CRP ticket will be generated but the generated CHANGELOG will not be consistent until I make that Stevenson PR)

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: If all other apps also wait for main Babylon app to be stable before creating a release branch, it would make more sense to put just `release/babylon/4.1.0` rather than `release/{appname}/{version}` to avoid confusion in future. We can mention somewhere in the document about other targets. By any means let's remove Telus from here because it is branched off `release/babylon/x.y.z` once it is stable and not off  `develop`

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (chitrakotwani) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: It might be worthwhile to add here that CHANGELOG will be created from difference of commits between two tags following the convention `target/x.y.z`. So that if some new engineer does not think of creating the tags(like me), is not left thinking why the CHANGELOG is not created

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (viorel15) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: What about release version and date ? Will that be added to `SDK/CHANGELOG.md` when this command is triggered?

**Review by viorel15 — APPROVED** [review]

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: the `SDK/CHANGELOG.md` is totally untouched by the CRP command, which is only based on the commit list, even for the SDK, and not the `CHANGELOG.md` file.

The `CHANGELOG.md` file will still need to be filled manually for now (since it's a file targeted for partners using the SDK, potentially worded in a more human-readable and filled manually). Besides, it should only be updated with the release date once the release is ready, so once QA has signed off the release, not at the time the release branch is cut, I guess, in case a change had to be introduced during a fix for a bug reported by QA.

What we can imagine doing though is have a `Danger` that will check that if the PR is for merging a branch named `release/sdk/{version}`, ensure that you have a `## {version}` section in the `SDK/CHANGELOG.md` file in that PR. That way we could check those at the time we're ready to merge the release (and thus the date should match), as opposed to do that when we cut off the release/sdk/{version} branch at the beginning of the release process. wdyt?

Comment (viorel15) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: Oh . I just noticed that the step was not removed. It's listed below. Thanks

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (Ben-Henshall) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: This is normally done by QA. Has this responsibility changed>

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (Ben-Henshall) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: This should be indented I think, since the numbering changes as a result

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (Ben-Henshall) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: These should be indented to prevent the numbering from being incorrect

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (Ben-Henshall) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: This will be fairly important imo, since there are quite a few SDK changes that are done by the individual squads rather than the SDK squad itself

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (Ben-Henshall) [inline Cookbook/Technical-Documents/ReleaseProcess.md]: Numbering also changes here and line 92

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: No idea. This line was there before my PR.
But indeed in the confluence doc it's mentioned QA is usually the one doing it, so I'll update accordingly.

File: Cookbook/Technical-Documents/ReleaseProcess.md
Comment (AliSoftware) [reply Cookbook/Technical-Documents/ReleaseProcess.md]: Nice catch!

**Review by Ben-Henshall — APPROVED** [review]

**Review by diegopetrucci — APPROVED** [review]

**Review by chitrakotwani — APPROVED** [review]

**Comment by AliSoftware** [comment]
Note: this new process is pending until [this PR for Stevenson is merged](https://github.com/Babylonpartners/ios-build-distribution/pull/67)

There are also 2 other tickets related to this new process:

* [This PR in Steve](https://github.com/Babylonpartners/ios-build-distribution/pull/67) is an optional change – it could allow us to still support the new naming in the legacy `/release` command, although at that point `/release` should not be used anymore (replaced by `/testflight` which the Stevenson PR intends to fix) so this PR is not really mandatory
* An upcoming PR (WIP) on the `Dangerfile` to force the use of `#SDK` in the PR title on any PR touching any file in `SDK/`

**Pull request merged**
